### DISCUSSION TITLE
Redis task handler for logs

### DISCUSSION THREAD

**Discussion by michalc** [apache/airflow#31832] [Ideas]
Should something like the below be in the codebase? It's a simple handler for storing Airflow task logs in Redis, enforcing a max number of entries per try, and an expiry time for the logs

Happy to raise a PR (and I guessed a lot at how things should be... so suspect can be improved upon...)

```python
class RedisHandler(logging.Handler):
    def __init__(self, client, key):
        super().__init__()
        self.client = client
        self.key = key

    def emit(self, record):
        p = self.client.pipeline()
        p.rpush(self.key, self.format(record))
        p.ltrim(self.key, start=-10000, end=-1)
        p.expire(self.key, time=60 * 60 * 24 * 28)
        p.execute()


class RedisTaskHandler(FileTaskHandler, LoggingMixin):
    """
    RedisTaskHandler is a python log handler that handles and reads
    task instance logs. It extends airflow FileTaskHandler and
    uploads to and reads from Redis.
    """

    trigger_should_wrap = True

    def __init__(self, base_log_folder: str, redis_url):
        super().__init__(base_log_folder)
        self.handler = None
        self.client = redis.Redis.from_url(redis_url)

    def _read(
        self,
        ti,
        try_number,
        metadata=None,
    ):
        log_str = b"\n".join(
            self.client.lrange(self._render_filename(ti, try_number), start=0, end=-1)
        ).decode("utf-8")
        return log_str, {"end_of_log": True}

    def set_context(self, ti):
        super().set_context(ti)
        self.handler = RedisHandler(
            self.client, self._render_filename(ti, ti.try_number)
        )
        self.handler.setFormatter(self.formatter)
```

**Comment by hussein-awala** [comment]
Considering Airflow's Redis provider, implementing a `TaskHandler` based on Redis should be ok, despite it not being commonly used as a log backend. If you believe it would be beneficial, I suggest raising a PR to add your classes.
However, please ensure that you update the authentication part using the Redis hook.

**Reply by michalc** [reply]
Have added a very WIP PR at https://github.com/apache/airflow/pull/31855. It does use a hook, but it has a customisable conn_id for it. For our case, we do have multiple logging handlers (for now at least, suspect not the long term plan), and so having one conn_id for them all is... tricky.
